File: CGConvertor/PresetsManager.swift
import Foundation

/// Presets Manager (Faza 1 v3.0.0, secțiunea D) — înlocuiește picker-ul
/// fix Mod(Rewrap/Transcode)+Codec cu o listă de presetări denumite,
/// editabile, importabile/exportabile. Persistat în `~/Library/
/// Application Support/CGConvertor/presets.json` — ACELAȘI nume de
/// fișier și STRUCTURĂ de câmpuri (`CodingKeys` mapate explicit la
/// snake_case) ca varianta Windows (`presets_manager.py`), ca un preset
/// exportat pe o platformă să se importe corect pe cealaltă.

enum AudioMode: String, CaseIterable, Identifiable, Codable {
    case passthrough, pcm16, pcm24, aac
    var id: String { rawValue }
    var labelKey: String {
        switch self {
        case .passthrough: return "audio.passthrough"
        case .pcm16: return "audio.pcm16"
        case .pcm24: return "audio.pcm24"
        case .aac: return "audio.aac"
        }
    }
    var ffmpegArgs: [String] {
        switch self {
        case .passthrough: return ["-c:a", "copy"]
        case .pcm16: return ["-c:a", "pcm_s16le"]
        case .pcm24: return ["-c:a", "pcm_s24le"]
        case .aac: return ["-c:a", "aac", "-b:a", "320k"]
        }
    }
}

enum ChannelLayout: String, CaseIterable, Identifiable, Codable {
    case original, stereo
    case surround51 = "5.1"
    var id: String { rawValue }
    var labelKey: String {
        switch self {
        case .original: return "channel.original"
        case .stereo: return "channel.stereo"
        case .surround51: return "channel.51"
        }
    }
    var ffmpegArgs: [String] {
        switch self {
        case .original: return []
        case .stereo: return ["-ac", "2"]
        case .surround51: return ["-ac", "6"]
        }
    }
}

/// Etichetare corectă a spațiului de culoare la ieșire (2026-09-05, cerut
/// explicit de Cristi — scop confirmat înainte de start: DOAR metadata
/// corectă în container/VUI, NU o transformare reală a pixelilor gen
/// LOG→Rec.709 sau Rec.709→Rec.2020, care ar cere filtrul `colorspace` al
/// FFmpeg cu risc real de artefacte pe combinații netestate — decizie
/// explicită de scop, nu o omisiune). `nil` (implicit) nu forțează nimic,
/// exact comportamentul de dinainte.
enum ColorSpaceOption: String, CaseIterable, Identifiable, Codable {
    case bt709, bt2020
    var id: String { rawValue }
    var labelKey: String {
        switch self {
        case .bt709: return "colorSpace.bt709"
        case .bt2020: return "colorSpace.bt2020"
        }
    }
    /// `-color_primaries`/`-color_trc`/`-colorspace` ca opțiuni brute de
    /// ieșire NU funcționează cu fiabilitate (verificat REAL: setează doar
    /// `colorspace`, nu și `primaries`/`transfer`, pe libx264 ȘI pe
    /// h264_videotoolbox) — filtrul `setparams` e metoda corectă, testată,
    /// de a scrie toate cele 3 etichete în VUI/container.
    var ffmpegArgs: [String] {
        switch self {
        case .bt709: return ["-vf", "setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709"]
        case .bt2020: return ["-vf", "setparams=color_primaries=bt2020:color_trc=bt2020-10:colorspace=bt2020nc"]
        }
    }
}

enum TargetApp: String, CaseIterable, Identifiable, Codable {
    case davinci, premiere, fcp, avid, web, custom
    var id: String { rawValue }
    var labelKey: String { "targetApp.\(rawValue)" }
}

struct OutputPreset: Identifiable, Codable, Equatable {
    var id: String
    var label: String
    var targetApp: TargetApp
    var profileID: String  // id din FormatRegistry, sau FormatRegistry.rewrapProfileID
    var audioMode: AudioMode = .passthrough
    var channelLayout: ChannelLayout = .original
    var fileSuffix: String = "_convertit"
    var isBuiltin: Bool = false
    /// Cadre/s la ieșire (2026-09-05, cerut explicit de Cristi) — `nil`
    /// (implicit) păstrează fps-ul sursei, exact comportamentul de dinainte
    /// de această schimbare (retrocompatibil 100% — presetările deja
    /// salvate pe disc, fără această cheie, decodează la `nil` automat,
    /// Optional-ul Swift tratează cheia lipsă ca `nil` fără decoder custom).
    /// Se aplică DOAR la transcodare — Rewrap (`-c copy`) nu poate resample
    /// fps fără re-encode, nu are sens acolo.
    var frameRate: String? = nil
    /// Etichetare spațiu de culoare la ieșire (2026-09-05) — `nil`
    /// (implicit) nu forțează niciun tag, comportamentul de dinainte.
    var colorSpace: ColorSpaceOption? = nil

    // Mapare explicită la snake_case — ACEEAȘI cheie JSON ca
    // `presets_manager.py` (Windows), pentru portabilitate reală
    // Import/Export între platforme, nu doar coincidență de nume.
    enum CodingKeys: String, CodingKey {
        case id, label
        case targetApp = "target_app"
        case profileID = "profile_id"
        case audioMode = "audio_mode"
        case channelLayout = "channel_layout"
        case fileSuffix = "file_suffix"
        case isBuiltin = "is_builtin"
        case frameRate = "frame_rate"
        case colorSpace = "color_space"
    }
}

/// Valorile de cadre/s oferite în picker — cele mai comune rate din
/// producție video (cinema/broadcast NTSC-PAL/web) — nu o listă liberă de
/// text, ca să nu se poată introduce o valoare invalidă pentru `-r`.
enum FrameRateOption {
    static let allValues = ["23.976", "24", "25", "29.97", "30", "50", "59.94", "60"]
}

enum PresetsManager {
    private static var fileURL: URL? {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first?
            .appendingPathComponent("CGConvertor", isDirectory: true)
            .appendingPathComponent("presets.json")
    }

    static func defaultPresets() -> [OutputPreset] {
        [
            OutputPreset(id: "builtin_rewrap", label: "Rewrap Rapid", targetApp: .custom,
                         profileID: FormatRegistry.rewrapProfileID, fileSuffix: "_rewrap", isBuiltin: true),
            OutputPreset(id: "builtin_prores422hq", label: "ProRes 422 HQ (Mezzanine DaVinci/FCP)",
                         targetApp: .davinci, profileID: "prores422hq", fileSuffix: "_proresHQ", isBuiltin: true),
            OutputPreset(id: "builtin_dnxhrhq", label: "DNxHR HQ (Mezzanine Avid/Premiere)",
                         targetApp: .avid, profileID: "dnxhrhq", fileSuffix: "_dnxhr", isBuiltin: true),
            OutputPreset(id: "builtin_h264_web", label: "H.264 1080p (YouTube/Web)",
                         targetApp: .web, profileID: "h264", audioMode: .aac, channelLayout: .stereo,
                         fileSuffix: "_web", isBuiltin: true),
            OutputPreset(id: "builtin_hevc_master", label: "HEVC 10-bit (Master Delivery)",
                         targetApp: .custom, profileID: "hevc10", audioMode: .aac, fileSuffix: "_master", isBuiltin: true),
            OutputPreset(id: "builtin_av1_web", label: "AV1 (Web modern)",
                         targetApp: .web, profileID: "av1", audioMode: .aac, channelLayout: .stereo,
                         fileSuffix: "_av1", isBuiltin: true),
            OutputPreset(id: "builtin_uncompressed", label: "Uncompressed 10-bit (Arhivare)",
                         targetApp: .custom, profileID: "uncompressed", audioMode: .pcm24,
                         fileSuffix: "_uncompressed", isBuiltin: true),
        ]
    }

    /// La prima lansare, seed cu presetările implicite (scrise pe disc,
    /// ca userul să le poată duplica/edita ca punct de plecare).
    static func load() -> [OutputPreset] {
        guard let url = fileURL,
              let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode([OutputPreset].self, from: data),
              !decoded.isEmpty else {
            let defaults = defaultPresets()
            save(defaults)
            return defaults
        }
        return decoded
    }

    static func save(_ presets: [OutputPreset]) {
        guard let url = fileURL else { return }
        try? FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        guard let data = try? JSONEncoder().encode(presets) else { return }
        try? data.write(to: url, options: .atomic)
    }

    static func duplicate(_ preset: OutputPreset, newID: String, newLabel: String) -> OutputPreset {
        var clone = preset
        clone.id = newID
        clone.label = newLabel
        clone.isBuiltin = false
        return clone
    }

    static func exportToFile(_ presets: [OutputPreset], url: URL) throws {
        let data = try JSONEncoder().encode(presets)
        try data.write(to: url, options: .atomic)
    }

    static func importFromFile(url: URL) throws -> [OutputPreset] {
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode([OutputPreset].self, from: data)
    }
}

File: python/presets_manager.py
"""
presets_manager.py — Presets Manager (CLAUDE.md, Faza 1 v3.0.0, sectiunea
D). Inlocuieste picker-ul fix Mod(Rewrap/Transcode)+Codec cu o lista de
presetari denumite, editabile, importabile/exportabile — fiecare preset
leaga un `EncoderProfile` (format_registry.py) de un mod audio si un
sufix de nume de fisier.

Persistat in acelasi folder ca `config.py` (`~/Library/Application
Support/CGConvertor/presets.json` pe Mac, `%APPDATA%\\CGConvertor\\
presets.json` pe Windows) — nu un fisier nou de configurare separat.
"""

import copy
import json
from dataclasses import asdict, dataclass
from typing import List

from config import config_dir  # acelasi folder, un singur loc care stie calea

PRESETS_PATH = config_dir() / "presets.json"

AUDIO_PASSTHROUGH = "passthrough"
AUDIO_PCM16 = "pcm16"
AUDIO_PCM24 = "pcm24"
AUDIO_AAC = "aac"

CHANNEL_ORIGINAL = "original"
CHANNEL_STEREO = "stereo"
CHANNEL_5_1 = "5.1"

_AUDIO_ARGS = {
    AUDIO_PASSTHROUGH: ["-c:a", "copy"],
    AUDIO_PCM16: ["-c:a", "pcm_s16le"],
    AUDIO_PCM24: ["-c:a", "pcm_s24le"],
    AUDIO_AAC: ["-c:a", "aac", "-b:a", "320k"],
}
_CHANNEL_ARGS = {
    CHANNEL_ORIGINAL: [],
    CHANNEL_STEREO: ["-ac", "2"],
    CHANNEL_5_1: ["-ac", "6"],
}


def audio_ffmpeg_args(audio_mode: str, channel_layout: str) -> List[str]:
    args = list(_AUDIO_ARGS.get(audio_mode, _AUDIO_ARGS[AUDIO_PASSTHROUGH]))
    if audio_mode != AUDIO_PASSTHROUGH:
        args += _CHANNEL_ARGS.get(channel_layout, [])
    return args


FRAME_RATE_OPTIONS = ["23.976", "24", "25", "29.97", "30", "50", "59.94", "60"]


@dataclass
class OutputPreset:
    id: str
    label: str
    target_app: str  # "davinci" / "premiere" / "fcp" / "avid" / "web" / "custom"
    profile_id: str  # id din format_registry.ALL_PROFILES, sau "rewrap"
    audio_mode: str = AUDIO_PASSTHROUGH
    channel_layout: str = CHANNEL_ORIGINAL
    file_suffix: str = "_convertit"
    is_builtin: bool = False  # seturile implicite (needitabile direct, dar duplicabile)
    # Cadre/s la iesire (2026-09-05, cerut explicit de Cristi) — None
    # (implicit) pastreaza fps-ul sursei, comportamentul de dinainte.
    # Retrocompatibil: presetarile deja salvate pe disc, fara aceasta
    # cheie, primesc None prin `from_dict` (filtreaza doar campurile
    # cunoscute, campurile lipsa raman pe valoarea implicita a dataclass-ului).
    # Se aplica DOAR la transcodare — Rewrap ("-c copy") nu poate resample
    # fps fara re-encode.
    frame_rate: str = None

    def to_dict(self):
        return asdict(self)

    @staticmethod
    def from_dict(data: dict) -> "OutputPreset":
        known = {f for f in OutputPreset.__dataclass_fields__}
        return OutputPreset(**{k: v for k, v in data.items() if k in known})


REWRAP_PROFILE_ID = "rewrap"  # nu e in format_registry — mod special, "-c copy" total


def _default_presets() -> List[OutputPreset]:
    return [
        OutputPreset(id="builtin_rewrap", label="Rewrap Rapid", target_app="custom",
                     profile_id=REWRAP_PROFILE_ID, audio_mode=AUDIO_PASSTHROUGH,
                     file_suffix="_rewrap", is_builtin=True),
        OutputPreset(id="builtin_prores422hq", label="ProRes 422 HQ (Mezzanine DaVinci/FCP)",
                     target_app="davinci", profile_id="prores422hq",
                     audio_mode=AUDIO_PASSTHROUGH, file_suffix="_proresHQ", is_builtin=True),
        OutputPreset(id="builtin_dnxhrhq", label="DNxHR HQ (Mezzanine Avid/Premiere)",
                     target_app="avid", profile_id="dnxhrhq",
                     audio_mode=AUDIO_PASSTHROUGH, file_suffix="_dnxhr", is_builtin=True),
        OutputPreset(id="builtin_h264_web", label="H.264 1080p (YouTube/Web)",
                     target_app="web", profile_id="h264",
                     audio_mode=AUDIO_AAC, channel_layout=CHANNEL_STEREO,
                     file_suffix="_web", is_builtin=True),
        OutputPreset(id="builtin_hevc_master", label="HEVC 10-bit (Master Delivery)",
                     target_app="custom", profile_id="hevc10",
                     audio_mode=AUDIO_AAC, channel_layout=CHANNEL_ORIGINAL,
                     file_suffix="_master", is_builtin=True),
        OutputPreset(id="builtin_av1_web", label="AV1 (Web modern)",
                     target_app="web", profile_id="av1",
                     audio_mode=AUDIO_AAC, channel_layout=CHANNEL_STEREO,
                     file_suffix="_av1", is_builtin=True),
        OutputPreset(id="builtin_uncompressed", label="Uncompressed 10-bit (Arhivare)",
                     target_app="custom", profile_id="uncompressed",
                     audio_mode=AUDIO_PCM24, file_suffix="_uncompressed", is_builtin=True),
    ]


def load() -> List[OutputPreset]:
    """La prima rulare, seed cu presetarile implicite (scrise pe disc, ca
    userul sa le poata duplica/edita ca punct de plecare). Presetarile
    corupte individual (JSON valid dar camp lipsa) sunt sarite, nu opresc
    incarcarea restului listei."""
    if not PRESETS_PATH.exists():
        defaults = _default_presets()
        save(defaults)
        return defaults
    try:
        with open(PRESETS_PATH, "r", encoding="utf-8") as f:
            raw = json.load(f)
        presets = []
        for item in raw:
            try:
                presets.append(OutputPreset.from_dict(item))
            except TypeError:
                continue
        return presets or _default_presets()
    except (json.JSONDecodeError, OSError):
        return _default_presets()


def save(presets: List[OutputPreset]) -> None:
    try:
        with open(PRESETS_PATH, "w", encoding="utf-8") as f:
            json.dump([p.to_dict() for p in presets], f, indent=2, ensure_ascii=False)
    except OSError:
        pass


def duplicate(preset: OutputPreset, new_id: str, new_label: str) -> OutputPreset:
    clone = copy.deepcopy(preset)
    clone.id = new_id
    clone.label = new_label
    clone.is_builtin = False
    return clone


def export_to_file(presets: List[OutputPreset], path: str) -> None:
    with open(path, "w", encoding="utf-8") as f:
        json.dump([p.to_dict() for p in presets], f, indent=2, ensure_ascii=False)


def import_from_file(path: str) -> List[OutputPreset]:
    """Intoarce presetarile importate — apelantul decide daca le adauga
    la lista existenta sau o inlocuieste. Nu suprascrie automat."""
    with open(path, "r", encoding="utf-8") as f:
        raw = json.load(f)
    result = []
    for item in raw:
        try:
            result.append(OutputPreset.from_dict(item))
        except TypeError:
            continue
    return result
